megatron/core/optimizer/layer_wise_optimizer.py

```diff
@@ -246,27 +262,39 @@ def sharded_state_dict(
                     0 if isinstance(sh_base.replica_id, int) else (*sh_base.replica_id[:2], 0)
                 )
 
+        # later code assume list but chained optimizer fallback to non-list if there's only one
         if len(self.chained_optimizers) == 1:
             wrapped_sharded_state_dict = {1: sharded_state_dict}
         else:
             wrapped_sharded_state_dict = sharded_state_dict
-        # Adjust dict due to possible empty rank 0 which output common_dict
+
+        # Adjust dict rank 0 output correct global metadata into common_dict
         for sd in wrapped_sharded_state_dict.values():
-            # Drop empty group state to avoid save in common dict (non-empty rank still save)
+            # wrap empty containers into LocalNonpersistentObject so it won't be saved/loaded
+            # params is already wrapped, we only need to handle fp32_from_fp16_params and state
+            # more details in load_state_dict comment
             if 'fp32_from_fp16_params' in sd:
                 sd['fp32_from_fp16_params'][:] = [
-                    group for group in sd['fp32_from_fp16_params'] if group
+                    group if group else LocalNonpersistentObject(group)
+                    for group in sd['fp32_from_fp16_params']
                 ]
-            # TODO(deyuf): 'common_step' code path is broken and 'step' is saved in 'param_groups'
-            # Find next 'step' if present. note this still break if rank0 adam is fully empty
-            step = next(
-                (group['step'] for group in sd['optimizer']['param_groups'] if 'step' in group),
-                None,
-            )
-            if step is not None:
-                for group in sd['optimizer']['param_groups']:
-                    group['step'] = step
-
+                sd['fp32_from_fp16_params'] = {
+                    k: v for k, v in enumerate(sd['fp32_from_fp16_params'])
```

Comment:
This is a bit confusing. I would use i instead of k here because it is not taking key from dict.item() iterator where  k,v is mostly commonly used.